src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -320,47 +316,42 @@ internal void MarkEntireType (TypeDefinition type, in DependencyInfo reason)
 
-			if (!_entireTypesMarked.Add (type))
-				return;
-
-			if (type.HasNestedTypes) {
-				foreach (TypeDefinition nested in type.NestedTypes)
-					MarkEntireType (nested, new DependencyInfo (DependencyKind.NestedType, type));
-			}
-
-			Annotations.Mark (type, reason);
-			MarkCustomAttributes (type, new DependencyInfo (DependencyKind.CustomAttribute, type));
-			MarkTypeSpecialCustomAttributes (type);
-
-			if (type.HasInterfaces) {
-				foreach (InterfaceImplementation iface in type.Interfaces)
-					MarkInterfaceImplementation (iface, new MessageOrigin (type));
+			// Calling MarkType here would result in extra warnings for cctors, ctors, delegate methods,
+			// serialization ctors, and overrides of preserved abstract methods.
+			// Instead, we only do extra type marking not already covered by the other cases in MarkEntireType.
+			void MarkTypeOnly (TypeDefinition type, in DependencyInfo reason)
+			{
+				Annotations.Mark (type, reason);
+				MarkCustomAttributes (type, new DependencyInfo (DependencyKind.CustomAttribute, type));
+				MarkTypeSpecialCustomAttributes (type);
+				MarkGenericParameterProvider (type);
 			}
 
-			MarkGenericParameterProvider (type);
+			MarkTypeOnly (type, reason);
 
-			if (type.HasFields) {
-				foreach (FieldDefinition field in type.Fields) {
-					MarkField (field, new DependencyInfo (DependencyKind.MemberOfType, type));
-				}
-			}
-
-			if (type.HasMethods) {
-				foreach (MethodDefinition method in type.Methods) {
+			// Note: This actually has nothing to do with dynamically accessed members - here we just use the DynamicallyAccessedMemberTypes.All
+			// as a convenient way to select every member of the type.
+			foreach (var member in type.GetDynamicallyAccessedMembers (_context, DynamicallyAccessedMemberTypes.All, declaredOnly: true)) {
+				switch (member) {
+				case MethodDefinition method:
 					Annotations.SetAction (method, MethodAction.ForceParse);
-					MarkMethod (method, new DependencyInfo (DependencyKind.MemberOfType, type));
-				}
-			}
-
-			if (type.HasProperties) {
-				foreach (var property in type.Properties) {
-					MarkProperty (property, new DependencyInfo (DependencyKind.MemberOfType, type));
-				}
-			}
-
-			if (type.HasEvents) {
-				foreach (var ev in type.Events) {
-					MarkEvent (ev, new DependencyInfo (DependencyKind.MemberOfType, type));
+					MarkMethod (method, reason);
+					break;
+				case FieldDefinition field:
+					MarkField (field, reason);
+					break;
+				case TypeDefinition nestedType:
+					MarkTypeOnly (nestedType, reason);
+					break;
+				case PropertyDefinition property:
+					MarkProperty (property, reason);
+					break;
+				case EventDefinition @event:
+					MarkEvent (@event, reason);
+					break;
+				case InterfaceImplementation iface:
+					MarkInterfaceImplementation (iface, new MessageOrigin (type));
+					break;
```

Comment:
If I read the code in `GetAllOnType` correctly, it will not return any interfaces when `declaredOnly:true`.

Means we're probably missing a test as well.